crates/wasi-common/src/sys/mod.rs

```diff
@@ -20,3 +23,60 @@ cfg_if! {
 
 pub(crate) use sys_impl::path;
 pub(crate) use sys_impl::poll;
+
+use super::handle::Handle;
+use crate::wasi::types;
+use osdir::OsDir;
+use osfile::OsFile;
+use osother::OsOther;
+use std::convert::TryFrom;
+use std::fs::File;
+use std::io;
+use std::mem::ManuallyDrop;
+use stdio::Stdio;
+use sys_impl::get_file_type;
+
+pub(crate) trait AsFile {
+    fn as_file(&self) -> ManuallyDrop<File>;
+}
+
+impl AsFile for dyn Handle + 'static {
```

Comment:
This is maybe a comment for the PR that introduced `Handle`, but if instead of having an `as_any(&self) -> &dyn Any`, the `Handle` trait instead depended on `Any`, then you could drop that trivial method, and also this `'static' constraint, because `Any` implies `'static`.

Reply:
Oh nice, I didn't think of that! Lemme give it a spin and report back. Oh, and thanks for the suggestion!

Reply:
OK, so after some careful analysis, I believe we'll still require `as_any(&self) -> &dyn Any` method defined in each implementor of the `Handle` trait. Here's the breakout of the reasons I discovered:
1. vtables in Rust currently do not permit for trait upcasting (see [here](https://articles.bchlr.de/traits-dynamic-dispatch-upcasting) and [here](https://github.com/rust-lang/rust/pull/60900)). This then implies that, even if `Handle: Any`, we can't go from `dyn Handle` to `dyn Any`.
2. Even if we added the constraint `Handle: Any` and made `as_any(&self) -> &dyn Any` a provided method, we'd still need to require that, at the very least, for `as_any(...)` `Self: Sized`. But then, `as_any(...)` is not available for `dyn Handle` trait object.

All in all, requiring `Any` on `Handle` seems to only get rid of the `'static` requirement in the referenced `impl AsFile` which I don't think is worth it? Lemme know what you think! Maybe there's something obvious that I missed and this will actually work, or perhaps I misunderstood what you meant! :-)

Reply:
I had no idea that trait upcasting was not allowed. That is a bummer and a sufficient reason for this design.

Reply:
I didn't either, so thanks for prodding me to have a look! :-)